### ISSUE TITLE
sessionresolver: run resolutions in parallel

### ISSUE THREAD

**Issue by bassosimone** [ooni/probe#1889]
We should measure, but intuitively I think we can afford 2-3 DoH resolvers in parallel. This would likely speed up startup.

**Comment by bassosimone** [comment]
I've started working to make this possible in https://github.com/ooni/probe-cli/pull/686. Before that refactoring, I tried a more direct approach, but it's difficult to do that in a single diff w/o writing not-so-good code because:

1. we're currently mutating the scores in place, which is of course not good when you have concurrency and predictably leads to data races when trying to write the data back;

2. if we do the naive thing to run three goroutines in parallel, the system resolver will be one of them many times and, of course, it will converge faster than the other resolvers, thus we'll basically end up using it most of the times, which is the fundamentally wrong thing to do (if that is fine, than we can just replace sessionresolver with a single function call that just invokes the system resolver and we'll significantly cut complexity, but of course this _is not_ fine);

3. another pain point is that I'd like to return as soon as possibly but I'd also like to defer writing back the status when the full new status is ready. I have yet to figure out how to avoid this pattern, but it's clear that there's need to refactor the way in which we use the disk in cache, to be more agile in this respect;

4. it may be that adding a constructor that reads the cache and changing the signature and name of the destructor to be like `Close` we may do wonders here: we may read the cache just once in the constructor and write it back on disk just once in the destructor. In the meanwhile, we can work on memory, so we can ignore the previous issue. (I am not super happy about renaming the destructor because this will mean that `Resolver` is not a `model.Resolver` but, on the other hand, it's clear that this specific of resolver needs to do something that is more `Close` than `CloseIdleConnections` and so it may be wise here to sacrifice a little purity and be a `Closer`).

5. it may be that treating uniformly all resolvers including the system resolver is wrong. If the system resolver was not part of the whole list, we could just run a bunch of DoH resolvers in parallel and then, if nothing happens for a couple of seconds or so, we can fire the system resolver. Yet, this design would mean we would take a timeout if none of the resolvers we support works and only the system resolver works. So, it may be that another strategy would be to just reorder the list such that the system resolver is tried after the first few best candidates. If we go down this road, we need  certainly to apply some changes to the way in which we add confusion.
